tests/naive.py

```diff
@@ -181,47 +182,83 @@ def stump(T_A, m, T_B=None, exclusion_zone=None):
     if exclusion_zone is None:
         exclusion_zone = int(np.ceil(m / config.STUMPY_EXCL_ZONE_DENOM))
 
-    if ignore_trivial:
-        diags = np.arange(exclusion_zone + 1, n_A - m + 1)
-    else:
-        diags = np.arange(-(n_A - m + 1) + 1, n_B - m + 1)
-
     P = np.full((l, 3), np.inf)
     I = np.full((l, 3), -1, dtype=np.int64)
 
-    for k in diags:
-        if k >= 0:
-            iter_range = range(0, min(n_A - m + 1, n_B - m + 1 - k))
-        else:
-            iter_range = range(-k, min(n_A - m + 1, n_B - m + 1 - k))
-
-        for i in iter_range:
-            D = distance_matrix[i, i + k]
-            if D < P[i, 0]:
-                P[i, 0] = D
-                I[i, 0] = i + k
+    if row_wise:  # row-wise traversal in distance matrix
+        if ignore_trivial:  # self-join
+            for i in range(l):
+                apply_exclusion_zone(distance_matrix[i], i, exclusion_zone, np.inf)
 
-            if ignore_trivial:  # Self-joins only
-                if D < P[i + k, 0]:
-                    P[i + k, 0] = D
-                    I[i + k, 0] = i
+        for i, D in enumerate(distance_matrix):
+            # self-join / AB-join: matrix proifle and indices
+            idx = np.argmin(D)
+            P[i, 0] = D[idx]
+            if P[i, 0] == np.inf:
+                idx = -1
+            I[i, 0] = idx
 
-                if i < i + k:
-                    # Left matrix profile and left matrix profile index
-                    if D < P[i + k, 1]:
-                        P[i + k, 1] = D
-                        I[i + k, 1] = i
-
-                    if D < P[i, 2]:
-                        # right matrix profile and right matrix profile index
-                        P[i, 2] = D
-                        I[i, 2] = i + k
-
-    result = np.empty((l, 4), dtype=object)
-    result[:, 0] = P[:, 0]
-    result[:, 1:4] = I[:, :]
-
-    return result
+            # self-join: left matrix profile
+            IL = -1
+            IR = -1
+            if ignore_trivial and i > 0:
+                IL = np.argmin(D[:i])
+                if D[IL] == np.inf:
+                    IL = -1
+                I[i, 1] = IL
+
+            # self-join: right matrix profile
+            if ignore_trivial and i < D.shape[0]:
+                IR = i + np.argmin(D[i:])  # shift argmin by `i` to get true index
+                if D[IR] == np.inf:
+                    IR = -1
+                I[i, 2] = IR
+
+        result = np.empty((l, 4), dtype=object)
+        result[:, 0] = P[:, 0]
+        result[:, 1:4] = I[:, :]
+
+        return result
```

Comment:
Don't return in the middle of the if statement. Please initialize `result` before the first (outermost) `if`, then fill it, and return `result` at the end of the function.

Reply:
Sure. I can do that. I will revise it.

Reply:
@seanlaw 

> Please initialize result before the first (outermost) if, then fill it, and return result at the end of the function.

Just out of curiosity, is there any particular reason for not doing all of that at the end of function? 

```
# in tests/naive.py
def stump(...):
    # at the end
    result = np.empty((l, 4), dtype=object)
    result[:, 0] = P[:, 0]
    result[:, 1:4] = I[:, :]
    
    return result
```

Reply:
Yes, at the end is fine too and is probably "better" in this case